include/bluetooth/gatt.h

```diff
@@ -1465,6 +1465,15 @@ struct bt_gatt_write_params {
  *  @param params Write parameters.
  *
  *  @return 0 in case of success or negative value in case of error.
+ *
+ *  @retval 0
+ *      Successfully queued request.
+ *      Will call @p params->func on resolution.
+ *
+ *  @retval -ENOMEM
+ *      Too many pending ATT requests.
+ *      Allow a pending request to resolve before retrying.
+ *      Relevant limit is @kconfig{CONFIG_BT_L2CAP_TX_BUF_COUNT}.
```

Comment:
This constitutes in some ways API breakage (in the sense that some application might have depended on the blocking behaviour for correct functionality). Would it make more sense to introduce a `bt_gett_write_async()` along side of the old API? If we don't do that we should follow the API change process and clearly document this in the release notes for Zephyr 3.0.

Reply:
I see your point. But just above in the function doc we have:

>  This procedure is asynchronous therefore the parameters need to remains valid while it is active.

So there was a mismatch between the documented API and behavior. This change makes the function adhere to its API. Given the function doc above, `bt_gatt_write_async` would be confusing.

Does this change your mind on this, @jhedberg?

Of course we should consider if we are breaking someones code. The blocking behavior could easily be relied on as flow control.

We could instead ratify the blocking behavior in the API. It may be possible to give a more informative LOG_ERR about the timeout. We could detect that we are on BT RX thread and report it as a deadlock. But what if we are in a more complex situation involving a transitive blocking where BT RX thread waits on different thread that is stuck in `bt_gatt_write`. Maybe we could detect a blocked BT RX thread on timeout and log it as the probable reason for the timeout. But it gets needlessly complex.

I prefer removing the blocking to remove any concerns of a deadlock. But I know how unpopular breaking other peoples code for the sake of correctness is. I would not be against scheduling this change for a major release.

Reply:
> Does this change your mind on this, @jhedberg?

Yes, that's a pretty good point. However, I still think it would be good to get maximal visibility on this change by having an entry in the release notes for it.

Reply:
This seems to be the same issue that we had with conn_tx_alloc.
Which we solved like this:
```
static struct bt_conn_tx *conn_tx_alloc(void)
{
	/* The TX context always get freed in the system workqueue,
	 * so if we're in the same workqueue but there are no immediate
	 * contexts available, there's no chance we'll get one by waiting.
	 */
	if (k_current_get() == &k_sys_work_q.thread) {
		return k_fifo_get(&free_tx, K_NO_WAIT);
	}
```
This is also documented in the API headers.

Not saying any solution is better than the other, but it gets confusing when it it is handled in multiple different ways.

Reply:
Good idea. `bt_gatt_write` is not the only public function that has changed. I think we should mention them all in the release notes. I'll have a look and find them. Should it go into doc/releases/release-notes-3.0.rst?

Reply:
> This procedure is asynchronous therefore the parameters need to remains valid while it is active.

@alwa-nordic This line got me thinking: When you read this, it sounds like procedure == function to you, but I am not sure I agree with that. The function may still be blocking, even if it an asynchronous procedure, so I don't agree that there is a mismatch between the documented API and the behavior. 

In any case, the blocking behavior of the function have caused significant amounts of issues when the receiver of callbacks in the RX thread attempts to send additional requests, so I am not against this change at all.

It will be different behavior from the POSIX `send()` function though. Not sure how badly Zephyr tries to be similar to POSIX.